Add File & Directory skipping

Twinki14 opened the pull request:
# Motivations
Sometimes there's files or directories that could be skipped during linking to overall speed up the process or give us an overall cleaner result. Recently such circumstances have included files suffixed with `.mohidden` or directories named `.git`

https://github.com/ModOrganizer2/usvfs/pull/60 is an example of an attempt at solving a specific example, but having such hardcoded values goes against what usvfs is

So a better solution is ideal

# Modifications
- Add two different lists with proper write & clear access functions that act very nearly identical to the existing executable blacklist, `skipFileSuffix` and `skipDirectories`.
- `skipFileSuffix` are suffixes that are compared to the filename during file linking, and if the ending of the filename matches one of the intended to-skip file suffixes, the file will be skipped
- `skipDirectory` acts similarly, but instead of a suffix the entire directory name is tested and compared

Twinki14 commented:
I'm still piecing together a PR for mod organizer that utilizes the newly added exposed functions, so this will stay as a draft in the meantime

Twinki14 commented:
Now ready for review

Holt59 commented on include/usvfs.h:
```diff
@@ -154,6 +154,30 @@ DLLEXPORT VOID WINAPI BlacklistExecutable(LPWSTR executableName);
+ * not to be confused with file extensions
+ * @param fileSuffix  a valid file suffix
+ */
+DLLEXPORT VOID WINAPI AddSkipFileSuffix(LPWSTR fileSuffix);
```
Although most other functions are not prefixed, I think it would be a good idea to prefix with `usvfs` similar to `usvfsConnectVFS` and `usvfsCreateVFS`.

Holt59 commented on .clang-format:
This should not be committed since the code is not actually formatted so this might just generates tons of changes when someone else start working on USVFS. If we want to have formatting here, that would be a separate PR.

Holt59 commented on include/usvfs.h:
```diff
@@ -154,6 +154,30 @@ DLLEXPORT VOID WINAPI BlacklistExecutable(LPWSTR executableName);
+
+/**
+ * adds a directory to be skipped during linking
+ * @param directory  full name of the directory
```
It's not clear at all, even after reading the whole PR, what `directory` represents here: a full absolute path? a directory name? a relative path to something?

Holt59 commented on src/usvfs_dll/hookcontext.cpp:
```diff
@@ -216,6 +216,56 @@ BOOL HookContext::executableBlacklisted(LPCWSTR wapp, LPCWSTR wcmd) const
+    return;
+  }
+
+  spdlog::get("usvfs")->debug("skipping files with the suffix '{}'", fsuffix);
```
This log (and the one for addSkipDirectory) is very misleading - We are not actually skipping files/directories here, we are registering them to be skipped.

Twinki14 replied:
I tried to follow the same way blacklisted execs are logged, I'll re-do these logs to give them more context

Holt59 commented on src/usvfs_dll/sharedparameters.cpp:
```diff
@@ -198,6 +200,82 @@ bool SharedParameters::executableBlacklisted(
+
+    for (const shared::StringT& skipFileSuffix : m_fileSuffixSkipList) {
+      if (boost::algorithm::iends_with(file, skipFileSuffix)) {
+        skipFile = true;
```
Just log and return `true` here (same applies for directory).

Holt59 commented on src/usvfs_dll/sharedparameters.cpp:
```diff
@@ -198,6 +200,82 @@ bool SharedParameters::executableBlacklisted(
+bool SharedParameters::fileShouldBeSkipped(const std::string& file) const
+{
+  bool skipFile = false;
+  std::string log;
```
This variable serves no purpose.

Holt59 commented on src/usvfs_dll/sharedparameters.cpp:
```diff
@@ -198,6 +200,82 @@ bool SharedParameters::executableBlacklisted(
+  {
+    bi::scoped_lock lock(m_mutex);
+
+    for (const shared::StringT& skipFileSuffix : m_fileSuffixSkipList) {
```
`const auto&`

Holt59 commented on src/usvfs_dll/sharedparameters.cpp:
```diff
@@ -198,6 +200,82 @@ bool SharedParameters::executableBlacklisted(
+  {
+    bi::scoped_lock lock(m_mutex);
+
+    for (const shared::StringT& skipDir : m_directorySkipList) {
```
`const auto&`

Holt59 commented on src/usvfs_dll/sharedparameters.cpp:
```diff
@@ -198,6 +200,82 @@ bool SharedParameters::executableBlacklisted(
+
+    for (const shared::StringT& skipDir : m_directorySkipList) {
+      if (boost::algorithm::equals(directory, skipDir)) {
+        skip = true;
```
Same as above, log and return here.

Holt59 commented on src/usvfs_dll/usvfs.cpp:
```diff
@@ -763,6 +763,11 @@ BOOL WINAPI VirtualLinkDirectoryStatic(LPCWSTR source, LPCWSTR destination, unsi
         if (file.attributes & FILE_ATTRIBUTE_DIRECTORY) {
           if ((file.fileName != L".") && (file.fileName != L"..")) {
+
+            if (context->directoryShouldBeSkipped(file.fileName)) {
```
Just add the negative of this to the `if` above, that would avoid an extra level and `if`.

Holt59 commented on src/usvfs_dll/usvfs.cpp:
```diff
@@ -771,6 +776,10 @@ BOOL WINAPI VirtualLinkDirectoryStatic(LPCWSTR source, LPCWSTR destination, unsi
           std::string nameU8 = ush::string_cast<std::string>(
               file.fileName.c_str(), ush::CodePage::UTF8);
 
+          if (context->fileShouldBeSkipped(file.fileName)) {
```
Put that as a `else if` instead of an inner `if`.

Twinki14 commented:
Updated, all requested changes have been addressed

Holt59 reviewed (approved):
Looks good to me, I'll wait a bit for @Silarn or @Al12rs review before merging.

Al12rs commented on src/usvfs_dll/hookcontext.cpp:
```diff
@@ -216,6 +216,56 @@ BOOL HookContext::executableBlacklisted(LPCWSTR wapp, LPCWSTR wcmd) const
+
+void usvfs::HookContext::addSkipDirectory(const std::wstring& directory) 
+{
+  const auto dir = shared::string_cast<std::string>(directory, shared::CodePage::UTF8);
```
Why do we convert everything to std::string? Could we keep everything in `std::wstring` ?
Haven't been dealing with C++ string types for a bit so not sure

Twinki14 replied:
I'm quite honestly not sure, I wanted to follow other usvfs code and assumed there's a reason for it. 

DLL exposed functions must take in a `wstring` or `LPCWSTR`, `LPCWSTR` is utf-16 and the gaps my brain has filled in is that we need to cast to utf-8 so that we can do system-compatible string comparisons (since not all systems will use utf-16 for filenames/directory names like Windows does)

Twinki14 replied:
Rubber ducking here, it looks like the file name pulled from the winapi that's passed into the file/directory functions to check if we should skip it all pass in a `wstring`, so what I said above is completely wrong.

`wstring` could be passed directly to all of these functions, but eventually there does need to be a cast to utf8 since the shared memory implementation of string is used heavily in usvfs, https://github.com/ModOrganizer2/usvfs/blob/master/src/shared/shared_memory.h#L50

Al12rs commented on src/usvfs_dll/sharedparameters.cpp:
```diff
@@ -198,6 +200,62 @@ bool SharedParameters::executableBlacklisted(
+
+bool SharedParameters::fileShouldBeSkipped(const std::string& file) const
+{
+  bi::scoped_lock lock(m_mutex);
```
We should be able to use a read write lock, like the shared_mutex, so we can have multiple readers at the same time.

JonathanAtCenterEdge replied:
That seems like a change for a different PR, this lock uses the same mutex used throughout this class

Twinki14 replied:
That seems out of scope for this PR, this lock uses a mutex used throughout this class

Al12rs replied:
I looked at the code again, so the problem is that we are calling those functions from a hot path, on every file that we add to the Redirect tree, which can be hundred of thousands of files. 

Every time we lock everything because we are accessing the shared parameters.
This isn't great, we should instead get the excluded lists once from shared memory, and then reuse that for all files, so we can bypass the lock. We don't expect the blacklist to be updated in the middle of the files being added, in if it was, we can ignore the changes to it since we already did a number of files with the previous filter anyways.

Do you think this proposed change is more doable?

Twinki14 replied:
That makes complete sense.

Instead of `file/directory..ShouldBeSkipped` I can add a simple `get` for each list, and call it in the `usvfs` class with a function that does what `..ShouldBeSkipped` essentially already does, but just takes the list from the previous `get` call.

I'll toy on this more when I find some time, very nice catch!

Twinki14 replied:
PR is ready for review again

Twinki14 commented on src/usvfs_dll/hookcontext.cpp:
```diff
@@ -216,6 +216,56 @@ BOOL HookContext::executableBlacklisted(LPCWSTR wapp, LPCWSTR wcmd) const
+
+BOOL usvfs::HookContext::fileShouldBeSkipped(const std::wstring& wFilename) const
+{
+  const std::string filename = ush::string_cast<std::string>(wFilename, ush::CodePage::UTF8);
```
Probably

Twinki14 commented:
PR is ready for review again

I've removed `fileShould..` functions and replaced them to simply return the list entirely of file suffixes / file directories and moved the logic to check the fileName against those lists inside the Virtual Linking functions.

Al12rs reviewed (changes requested):
Seems good, I reviewed from phone though.
Have you been able to test and verify things are being skipped correctly?

Twinki14 commented on include/usvfs.h:
```diff
@@ -49,6 +49,7 @@ static const unsigned int LINKFLAG_CREATETARGET   = 0x00000004; // if set, file
                                                                 // If there different create-target have been set for an element and one of its
                                                                 // ancestors, the inner-most create-target is used
 static const unsigned int LINKFLAG_RECURSIVE      = 0x00000008; // if set, directories are linked recursively
+static const unsigned int LINKFLAG_FAILIFSKIPPED  = 0x00000010; // if set, linking fails if the file or directory is skipped
```
> Seems good, I reviewed from phone though. Have you been able to test and verify things are being skipped correctly?

Yes, it's been fairly well test-covered through https://github.com/ModOrganizer2/modorganizer/pull/2033

> Comment should mention the skip lists to give contetext as to how a file might actually be skipped. Also put the comment on a separate line above, simply to avoid long lines.

Updated

Holt59 commented on src/usvfs_dll/sharedparameters.cpp:
```diff
@@ -198,6 +200,46 @@ bool SharedParameters::executableBlacklisted(
+  m_fileSuffixSkipList.clear();
+}
+
+std::vector<std::string> SharedParameters::skipFileSuffixes() const
```
I don't really understand why this returns a new vector - This seems pretty costly as this needs to create a vector (although of small size) for every linked file/directory.

Do you assume the underlying list could be modified while inside a call to VirtualLinkXXX?

Twinki14 replied:
It would create a new vector only once in `VirtualLinkDirectoryStatic` which from my understanding is intended to be used when you're dealing with one gigantic directory (like a Game)

`VirtualLinkFile` would create a new vector every time a **specific** file is linked, from my understanding `VirtualLinkFile` would be called if you only want to link ONE file externally. `VirtualLinkFile` isn't used in `VirtualLinkDirectoryStatic`

Twinki14 replied:
Also, yes I opted to copy the internal list into a new vector since this class could be accessed and used by simultaneous threads.
